File: solutions/javascript/largest-series-product/1/largest-series-product.js
/**
 * Calculates the largest product for a contiguous substring of digits of specified length.
 *
 * @param {string} digits - The string of digits to search.
 * @param {number} seriesLength - The length of the substring for which to calculate the product.
 * @returns {number} - The largest product found.
 * @throws {Error} - If the series length is negative, greater than the length of the digits, or if the digits string contains non-digit characters.
 */
export const largestProduct = (digits, seriesLength) => {
  if (seriesLength === 0) {
    return 1;
  }

  if (seriesLength > digits.length) {
    throw new Error('span must be smaller than string length');
  }

  if (seriesLength < 0) {
    throw new Error('span must not be negative');
  }

  if (!/^\d+$/g.test(digits)) {
    throw new Error('digits input must only contain digits');
  }

  let result = 0;

  for (let i = 0; i <= digits.length - seriesLength; i++) {
    const product = digits
      .substring(i, i + seriesLength)
      .split('')
      .map(Number)
      .reduce((a, b) => a * b, 1); // Provide initial value for reduce

    if (product > result) {
      result = product;
    }
  }

  return result;
};

File: solutions/python/largest-series-product/1/largest_series_product.py
from functools import reduce


def largest_product(series, size):
    if size == 0:
        return 1
    if size < 0:
        raise ValueError("span must not be negative")
    elif (series_length := len(series := str(series))) < size:
        raise ValueError("span must be smaller than string length")
    elif not series.isdigit():
        raise ValueError("digits input must only contain digits")

    return max(
        reduce(
            lambda x, y: x * y,
            tuple(int(n) for n in series)[index : index + size],
        )
        for index in range(series_length - (size - 1))
    )
